### PR TITLE
865- Events for challenge

### PR DESCRIPTION
**Pull request by akileshtangella**
- [ ] Set a valid title
- [ ] Use branch-name `[<ref>-]<feature-description>`
- [ ] Brief description of the changes
- [ ] Create matching PRs in dependent repos

### CODE REVIEW

File: upa/src/sdk/events.ts [hunk 1]
```diff
@@ -321,3 +323,90 @@ export class VKRegisteredEventGetter extends EventGetterBase<
...
+  const decoded = verifier.interface.decodeFunctionData(
+    challengeFragment,
+    tx.data
+  );
+
+  // The decoded data is dynamically indexed.  Extract everything to make it a
+  // concrete struct.
+
+  const circuitId: string = readBytes32(decoded.circuitId);
+  const proof: Groth16CompressedProofStruct = decoded.proof;
```
Comment (dtebbs) [inline upa/src/sdk/events.ts]: Is this correct?  Don't we need to do `from_solidity` or something?

File: upa/src/sdk/events.ts [hunk 2]
```diff
@@ -321,3 +323,90 @@ export class VKRegisteredEventGetter extends EventGetterBase<
...
+  const publicInputs: bigint[] = decoded.publicInputs.map(
+    (pi: ethers.BigNumberish) => BigInt(pi)
+  );
```
Comment (dtebbs) [inline upa/src/sdk/events.ts]: Might be able to just do:
```suggestion
  const publicInputs = decoded.publicInputs.map(BigInt);
```

File: upa/src/sdk/events.ts (hunk 1 above)
Comment (dtebbs) [reply upa/src/sdk/events.ts]: In fact, isn't this uncompressed?

**Review by dtebbs — COMMENTED** [review]
Partial review, but I just realised you didn't mark me.  Apologies.

**Comment by akileshtangella** [comment]
> Partial review, but I just realised you didn't mark me. Apologies.

No problem, thanks for the comment! Addressing them now. Was working on writing the `test_challenge` script.

File: upa/src/sdk/events.ts (hunk 1 above)
Comment (akileshtangella) [reply upa/src/sdk/events.ts]: Yes, fixed.

File: upa/src/sdk/events.ts (hunk 2 above)
Comment (akileshtangella) [reply upa/src/sdk/events.ts]: Done.

File: upa/scripts/test_challenge [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    # Deploy UPA using this key. We set a fixed fee per proof of 20 Wei.
+    KEYFILE_PASSWORD=${DUMMY_PW} upa owner deploy \
+                    --keyfile ${KEYFILE} \
+                    --verifier ${UPA_DIR}/upa/test/data/test.bin \
+                    --collateral 10000000000000000000 \
+                    --use-test-config \
+                    --retries 3
+
+    # Print the config
+    upa query config > config.upa.json
```
Comment (dtebbs) [inline upa/scripts/test_challenge]: Do we need this?

File: upa/scripts/test_challenge [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    "
+
+    # Submit the aggregated proof
+    KEYFILE_PASSWORD=${DUMMY_PW} upa aggregator submit-aggregated-proof \
+           --keyfile ${KEYFILE} \
+           ${submit_aggregated_proof_flags} \
+           --wait > agg-proof.hash
+
+    # Create a dump of the aggregated tx.
+    upa dev trace `cat agg-proof.hash` --trace-file agg-proof.trace.json
```
Comment (dtebbs) [inline upa/scripts/test_challenge]: What's this for?

**Review by dtebbs — APPROVED** [review]
Some nits, otherwise LGTM.  Great to have this working.  

This new test adds over a minute to the CI.  The main operation of the contract should be tested by the unit tests, so we don't need to go too deep in terms of checking everything.  Main point is to test that the tool works as expected.  Having said that, no problem if there is nothing obvious we can cull.

File: upa/scripts/test_challenge (hunk 4 above)
Comment (akileshtangella) [reply upa/scripts/test_challenge]: Removed. 

It is in `test_upa` script which I based this one off of. But doesn't seem to be used.

File: upa/scripts/test_challenge (hunk 3 above)
Comment (akileshtangella) [reply upa/scripts/test_challenge]: Rmoved, thanks.

**Pull request merged**